File: backend/src/models/Invoice.ts
import mongoose, { Schema, Document } from 'mongoose';
import { IUser } from './User';
import { IClient } from './Client';
import { IServicePlan } from './ServicePlan';

export interface IInvoice extends Document {
  invoiceNumber: string;
  ownerId: IUser['_id'];
  clientId?: IClient['_id'];
  subscriptionId?: mongoose.Types.ObjectId;
  planId?: IServicePlan['_id'];
  status: 'DRAFT' | 'SENT' | 'PAID' | 'OVERDUE' | 'CANCELLED';
  amountKes: number;
  vatRate: number;
  vatAmount: number;
  totalAmount: number;
  billingPeriod?: string;
  dueDate: Date;
  paidDate?: Date;
  mpesaReceiptNo?: string;
  notes?: string;
  createdAt: Date;
  updatedAt: Date;
}

const InvoiceSchema: Schema = new Schema({
  invoiceNumber: { type: String, required: true, unique: true },
  ownerId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  clientId: { type: Schema.Types.ObjectId, ref: 'Client' },
  subscriptionId: { type: Schema.Types.ObjectId, ref: 'Subscription' },
  planId: { type: Schema.Types.ObjectId, ref: 'ServicePlan' },
  status: { 
    type: String, 
    enum: ['DRAFT', 'SENT', 'PAID', 'OVERDUE', 'CANCELLED'], 
    default: 'DRAFT' 
  },
  amountKes: { type: Number, required: true },
  vatRate: { type: Number, default: 16 },
  vatAmount: { type: Number, default: 0 },
  totalAmount: { type: Number, required: true },
  billingPeriod: { type: String },
  dueDate: { type: Date, required: true },
  paidDate: { type: Date },
  mpesaReceiptNo: { type: String },
  notes: { type: String },
}, { timestamps: true });

InvoiceSchema.index({ ownerId: 1 });
InvoiceSchema.index({ invoiceNumber: 1 });
InvoiceSchema.index({ status: 1 });

export default mongoose.model<IInvoice>('Invoice', InvoiceSchema);

File: backend/app/models/invoice.py
from datetime import datetime, timezone
from typing import Optional, Literal
from beanie import Document, PydanticObjectId
from pydantic import Field


class Invoice(Document):
    invoice_number: str = Field(unique=True, alias="invoiceNumber")
    owner_id: PydanticObjectId = Field(alias="ownerId")
    client_id: Optional[PydanticObjectId] = Field(default=None, alias="clientId")
    subscription_id: Optional[PydanticObjectId] = Field(
        default=None, alias="subscriptionId"
    )
    plan_id: Optional[PydanticObjectId] = Field(default=None, alias="planId")
    status: Literal["DRAFT", "SENT", "PAID", "OVERDUE", "CANCELLED"] = "DRAFT"
    amount_kes: float = Field(alias="amountKes")
    vat_rate: float = Field(default=16, alias="vatRate")
    vat_amount: float = Field(default=0, alias="vatAmount")
    total_amount: float = Field(alias="totalAmount")
    billing_period: Optional[str] = Field(default=None, alias="billingPeriod")
    due_date: datetime = Field(alias="dueDate")
    paid_date: Optional[datetime] = Field(default=None, alias="paidDate")
    mpesa_receipt_no: Optional[str] = Field(default=None, alias="mpesaReceiptNo")
    notes: Optional[str] = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "invoices"
        populate_by_name = True
        indexes = [
            "owner_id",
            "invoice_number",
            "status",
        ]
